### PR TITLE
Refactor of GLPK solvers

### LINKED ISSUE #760: Refactor GLPK solvers
We have two LP implementations using GLPK. There is a lot of replicated code in both implementations. A refactor that puts the boiler plate (e.g. basis management, init, etc.) into a shared base class would improve maintainability and perhaps open up possibilities for making more LP implementations (e.g. MILP #702).

### PR DESCRIPTION
**Pull request by snorfalorpagus**
Refactor of common code in the GLPK solvers into a shared parent class.

Closes #760.

### CODE REVIEW

File: pywr/solvers/cython_glpk_edge.pyx [hunk 1]
```diff
@@ -6,24 +6,21 @@ cimport numpy as np
...
 cimport cython
 
+from .glpk cimport GLPKSolver, BasisManager, simplex, set_obj_coef, set_row_bnds, set_col_bnds, set_mat_row
+
 from pywr._core import BaseInput, BaseOutput, BaseLink
 from pywr._core cimport *
 from pywr.core import ModelStructureError
 import time
 
 include "glpk.pxi"
```
Comment (jetuk) [inline pywr/solvers/cython_glpk_edge.pyx]: Is there a reason to keep `glpk.pxi` at all? Why can't all the function definitions be in the `glpk.pxd`?

File: pywr/solvers/cython_glpk.pyx [hunk 2]
```diff
@@ -577,15 +518,8 @@ cdef class CythonGLPKSolver:
...
-        # Apply presolve if required
-        if self.use_presolve and not self.has_presolved:
-            self.smcp.presolve = GLP_ON
-            self.has_presolved = True
-        else:
-            self.smcp.presolve = GLP_OFF
-
         # Set the basis for this scenario
-        self._set_basis(scenario_index.global_id)
+        self.basis_manager.set_basis(self.prob, self.is_first_solve, scenario_index.global_id)
```
Comment (jetuk) [inline pywr/solvers/cython_glpk.pyx]: Can we abstract the whole section on solving the model and error handling (522-549 here) into the base class?

File: pywr/solvers/cython_glpk.pyx [hunk 3]
```diff
@@ -395,48 +372,12 @@ cdef class CythonGLPKSolver:
...
-            for i in range(nrows):
-                glp_set_row_stat(self.prob, i+1, self.row_stat[global_id, i])
-            for i in range(ncols):
-                glp_set_col_stat(self.prob, i+1, self.col_stat[global_id, i])
-
     def reset(self):
         # Resetting this triggers a crashing of a new basis in each scenario
         self.is_first_solve = True
 
     cpdef object solve(self, model):
```
Comment (jetuk) [inline pywr/solvers/cython_glpk.pyx]: Can this be done in the base class? Looks the same between the two implementations?

File: pywr/solvers/cython_glpk.pyx [hunk 4]
```diff
@@ -395,48 +372,12 @@ cdef class CythonGLPKSolver:
...
-            # otherwise we restore basis from previous solve of this scenario
-            nrows = glp_get_num_rows(self.prob)
-            ncols = glp_get_num_cols(self.prob)
-
-            for i in range(nrows):
-                glp_set_row_stat(self.prob, i+1, self.row_stat[global_id, i])
-            for i in range(ncols):
-                glp_set_col_stat(self.prob, i+1, self.col_stat[global_id, i])
-
     def reset(self):
```
Comment (jetuk) [inline pywr/solvers/cython_glpk.pyx]: Can this be done in the base class? Looks the same between the two implementations?

**Comment by snorfalorpagus** [comment]
The choice to leave the `reset` and `solve` methods in the child classes was a deliberate one. Neither of those functions are very long, so the duplication isn't too painful. These methods are identical in the two existing solvers but I can imagine others where they are different. For example, a solver than handles multiple scenarios in the same LP. The choice to keep `glpk.pxi` separate is for the same reason - perhaps someone would write a solver that works completely differently but still wants to use GLPK. Also it needs to be in a pxi rather than pxd because there are a couple of helper functions in there still.

**Pull request merged**